File: Frontend/app/src/main/java/com/example/CyTrack/Startup/LoginActivity.java
package com.example.CyTrack.Startup;

import android.content.Intent;
import android.os.Bundle;
import android.widget.EditText;

import androidx.activity.EdgeToEdge;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;
import androidx.core.splashscreen.SplashScreen;

import android.widget.Button;

import android.widget.Toast;

import com.example.CyTrack.Dashboard.MainDashboardActivity;
import com.example.CyTrack.R;
import com.example.CyTrack.Utilities.UrlHolder;
import com.example.CyTrack.Utilities.User;
import com.example.CyTrack.Utilities.FocusUtils;
import com.example.CyTrack.Utilities.NetworkUtils;

import java.util.HashMap;
import java.util.Map;


/**
 * LoginActivity handles the login process for the application.
 */
public class LoginActivity extends AppCompatActivity {

    /**
     * EditText field for the username input.
     */
    private EditText usernameEditText;

    /**
     * EditText field for the password input.
     */
    private EditText passwordEditText;

    /**
     * Button to navigate to the sign-up activity.
     */
    private Button signUpButton;

    /**
     * Button to initiate the login process.
     */
    private Button loginButton;

    /**
     * Button to navigate to the forgot password activity.
     */
    private Button forgotPasswordButton;

    /**
     * User object to store the logged-in user's information.
     */
    private User user;

    /**
     * URL for the login endpoint.
     */
    private final String URL_LOGIN = UrlHolder.URL + "/user/login";

    /**
     * URL for fetching user data.
     */
    private final String URL_GET_USER = UrlHolder.URL + "/user";

    /**
     * Called when the activity is first created.
     *
     * @param savedInstanceState If the activity is being re-initialized after previously being shut down then this Bundle contains the data it most recently supplied in onSaveInstanceState(Bundle).
     */
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        SplashScreen splashScreen = SplashScreen.installSplashScreen(this);
        super.onCreate(savedInstanceState);
        EdgeToEdge.enable(this);
        setContentView(R.layout.startup_login_activity);
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main), (v, insets) -> {
            Insets systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom);
            return insets;
        });

        usernameEditText = findViewById(R.id.Username_field);
        passwordEditText = findViewById(R.id.Password_field);
        signUpButton = findViewById(R.id.signUp_button);
        loginButton = findViewById(R.id.login_button);
        forgotPasswordButton = findViewById(R.id.forgotPassword_button);

        loginButton.setOnClickListener(v -> {
            String username = usernameEditText.getText().toString();
            String password = passwordEditText.getText().toString();

            if (username.isEmpty() || password.isEmpty()) {
                Toast.makeText(getApplicationContext(), "Please fill in all fields", Toast.LENGTH_LONG).show();
                return;
            }

            login(username, password);
        });

        signUpButton.setOnClickListener(v -> {
            Intent intent = new Intent(LoginActivity.this, SignUpActivity.class);
            startActivity(intent);
        });

        forgotPasswordButton.setOnClickListener(v -> {
            Intent intent = new Intent(LoginActivity.this, ForgotPasswordActivity.class);
            startActivity(intent);
        });

        clearHintOnFocus();
    }

    /**
     * Initiates the login process by sending the username and password to the server.
     *
     * @param username The username entered by the user.
     * @param password The password entered by the user.
     */
    private void login(String username, String password) {
        Map<String, String> params = new HashMap<>();
        params.put("username", username);
        params.put("password", password);

        NetworkUtils.postUserAndGetID(getApplicationContext(), URL_LOGIN, params, new NetworkUtils.postUserAndGetIDCallback() {
            @Override
            public void onSuccess(int id, String message) {
                if (id != 0) fetchUserData(URL_GET_USER + "/" + id);
            }

            @Override
            public void onError(String error) {
                Toast.makeText(getApplicationContext(), error, Toast.LENGTH_LONG).show();
            }
        });
    }

    /**
     * Fetches user data from the server using the provided URL.
     *
     * @param url The URL to fetch user data from.
     */
    private void fetchUserData(String url) {
        NetworkUtils.fetchUserData(this, url, new NetworkUtils.fetchUserDataCallback() {
            @Override
            public void onSuccess(User user, String message) {
                LoginActivity.this.user = user;
                Toast.makeText(getApplicationContext(), "Signing In", Toast.LENGTH_LONG).show();
                navigateToMainDashboard();
            }

            @Override
            public void onError(String error) {
                Toast.makeText(getApplicationContext(), error, Toast.LENGTH_LONG).show();
            }
        });
    }

    /**
     * Navigates to the main dashboard activity.
     */
    private void navigateToMainDashboard() {
        Intent intent = new Intent(LoginActivity.this, MainDashboardActivity.class);
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
        intent.putExtra("user", user);
        startActivity(intent);
    }

    /**
     * Clears the hint text when the EditText fields gain focus.
     */
    private void clearHintOnFocus() {
        FocusUtils.clearHintOnFocus(usernameEditText, "Username");
        FocusUtils.clearHintOnFocus(passwordEditText, "Password");
    }
}

File: Frontend/app/src/main/java/com/example/CyTrack/Startup/Login/LoginActivity.kt
package com.example.CyTrack.Startup.Login

import android.content.Intent
import android.os.Bundle
import android.widget.Toast
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.size
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Surface
import androidx.compose.runtime.MutableState
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.unit.dp
import com.example.CyTrack.Dashboard.MainDashboardActivity
import com.example.CyTrack.R
import com.example.CyTrack.Startup.ForgotPassword.Activities.CheckUsernameActivity
import com.example.CyTrack.Startup.ForgotPasswordActivity
import com.example.CyTrack.Startup.SignUp.Activities.UsernameActivity
import com.example.CyTrack.Startup.SignUp.Composables.BasicRedCircularLoadingScreen
import com.example.CyTrack.Utilities.NetworkUtils
import com.example.CyTrack.Utilities.NetworkUtils.fetchUserDataCallback
import com.example.CyTrack.Utilities.NetworkUtils.postUserAndGetIDCallback
import com.example.CyTrack.Utilities.StatusBarUtil
import com.example.CyTrack.Utilities.UrlHolder
import com.example.CyTrack.Utilities.User
import com.example.compose.CyRedMain

class LoginActivity : ComponentActivity() {

    private lateinit var user: User

    private lateinit var username: MutableState<String>

    private lateinit var password: MutableState<String>

    private lateinit var loading: MutableState<Boolean>

    private val URL_LOGIN = "${UrlHolder.URL}/user/login"

    private val URL_GET_USER = "${UrlHolder.URL}/user"

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        StatusBarUtil.setStatusBarColor(this, R.color.white)
        setContent {
            username = remember { mutableStateOf("") }
            password = remember { mutableStateOf("") }
            loading = remember { mutableStateOf(false) }

            if (loading.value) {
                BasicRedCircularLoadingScreen(loading)
            }

            Surface(
                modifier = Modifier
                    .fillMaxSize()
                    .then(if (loading.value) Modifier.alpha(0.5f) else Modifier)
            ) {
                LoginScreen(
                    username = username,
                    password = password,
                    onLogin = {
                        login(username.value, password.value)
                        loading.value = true
                    },
                    onForgotPassword = {
                        navigateToForgotPassword()
                    },
                    onSignUp = {
                        navigateToSignUp()
                    }
                )
            }

        }

    }

    /**
     * Initiates the login process by sending the username and password to the server.
     *
     * @param username The username entered by the user.
     * @param password The password entered by the user.
     */
    private fun login(username: String, password: String) {
        val params: MutableMap<String, String> = HashMap()
        params["username"] = username
        params["password"] = password

        NetworkUtils.postUserAndGetID(
            applicationContext,
            URL_LOGIN,
            params,
            object : postUserAndGetIDCallback {
                override fun onSuccess(id: Int, message: String) {
                    if (id != 0) fetchUserData(URL_GET_USER + "/" + id)
                }

                override fun onError(error: String) {
                    loading.value = false
                    Toast.makeText(applicationContext, error, Toast.LENGTH_LONG).show()
                }
            })
    }

    private fun fetchUserData(url: String) {
        NetworkUtils.fetchUserData(this, url, object : fetchUserDataCallback {
            override fun onSuccess(user: User, message: String) {
                this@LoginActivity.user = user
                Toast.makeText(applicationContext, "Signing In", Toast.LENGTH_LONG).show()
                loading.value = false
                navigateToMainDashboard()
            }

            override fun onError(error: String) {
                loading.value = false
                Toast.makeText(applicationContext, error, Toast.LENGTH_LONG).show()
            }
        })
    }

    private fun navigateToMainDashboard() {
        val intent = Intent(
            this@LoginActivity,
            MainDashboardActivity::class.java
        )
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK)
        intent.putExtra("user", user)
        startActivity(intent)
    }

    private fun navigateToForgotPassword() {
        val intent = Intent(
            this@LoginActivity,
            CheckUsernameActivity::class.java
        )
        startActivity(intent)
    }

    private fun navigateToSignUp() {

        val intent = Intent(
            this@LoginActivity,
            UsernameActivity::class.java
        )
        startActivity(intent)
    }
}
